### PR TITLE
feat:Use shardmap to optimize healthcheck performance

### PR DESCRIPTION
**Pull request by liu-song**
**Please provide issue(s) of this PR:**
Fixes https://github.com/polarismesh/polaris/issues/205

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration
- [ ] Docs
- [ ] Installation
- [x] Performance and Scalability
- [ ] Naming
- [x] HealthCheck
- [ ] Test and Release

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any user-facing changes. This may include API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: service/healthcheck/cache.go [hunk 1]
```diff
@@ -18,20 +18,25 @@
...
 	api "github.com/polarismesh/polaris-server/common/api/v1"
 	"github.com/polarismesh/polaris-server/common/model"
 	"github.com/polarismesh/polaris-server/plugin"
+	"runtime"
 )
 
+var ShardSize uint32
+
+func init() {
+	ShardSize = uint32(runtime.GOMAXPROCS(0) * 16)
```
Comment (chuntaojun) [inline service/healthcheck/cache.go:30]: ShardSize => DefaultShardSize

File: service/healthcheck/cache.go [hunk 2]
```diff
@@ -62,17 +65,15 @@ func (c *CacheProvider) sendEvent(event CacheEvent) {
...
 }
 
-func compareAndStoreServiceInstance(
-	instanceWithChecker *InstanceWithChecker, mutex *sync.RWMutex, values map[string]*InstanceWithChecker) bool {
-	mutex.Lock()
-	defer mutex.Unlock()
+func compareAndStoreServiceInstance(instanceWithChecker *InstanceWithChecker, values *ShardMap) bool {
 	instanceId := instanceWithChecker.instance.ID()
-	value, ok := values[instanceId]
+	value, ok := values.Load(instanceId)
```
Comment (chuntaojun) [inline service/healthcheck/cache.go:70]: 是这里，这个应该直接采用 computeIfAbsent 的模式，返回 value，isNew
如果存在，返回旧的value，isNew=false，否则返回计算出来的value，isNew=true
然后在接着走下面的逻辑

File: service/healthcheck/cache.go [hunk 3]
```diff
@@ -81,31 +82,27 @@ func compareAndStoreServiceInstance(
...
 	return true
 }
 
-func deleteServiceInstance(instance *api.Instance, mutex *sync.RWMutex, values map[string]*InstanceWithChecker) bool {
-	mutex.Lock()
-	defer mutex.Unlock()
+func deleteServiceInstance(instance *api.Instance, values *ShardMap) bool {
 	instanceId := instance.GetId().GetValue()
-	_, ok := values[instanceId]
+	_, ok := values.Load(instanceId)
```
Comment (chuntaojun) [inline service/healthcheck/cache.go:101]: can use RemoveIfAbsent，降低整体锁的获取次数

File: service/healthcheck/cache.go [hunk 4]
```diff
@@ -230,27 +227,24 @@ func (c *CacheProvider) OnBatchDeleted(value interface{}) {
...
 
 // RangeSelfServiceInstances range loop selfServiceInstances
 func (c *CacheProvider) RangeSelfServiceInstances(check func(instance *api.Instance)) {
-	c.selfServiceMutex.RLock()
-	defer c.selfServiceMutex.RUnlock()
-	for _, value := range c.selfServiceInstances {
-		check(value.instance.Proto)
-	}
+
+	c.selfServiceInstances.RangeMap(func(instanceId string, selfServiceInstance *InstanceWithChecker) {
```
Comment (chuntaojun) [inline service/healthcheck/cache.go:239]: RangeMap => Range

**Comment by liu-song** [comment]
当前版本，能准确计算出 shardmap 中元素个数的benchmark:
goos: darwin
goarch: amd64
pkg: github.com/polarismesh/polaris-server/service/healthcheck
cpu: Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
BenchmarkSyncMap
BenchmarkSyncMap-8               1000436              1075 ns/op             134 B/op          4 allocs/op
BenchmarkMutexMap
BenchmarkMutexMap-8              2643946               485.8 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap2
BenchmarkSharedMap2-8            2458927               453.1 ns/op             9 B/op          2 allocs/op
BenchmarkSharedMap4
BenchmarkSharedMap4-8            2950054               389.8 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap8
BenchmarkSharedMap8-8            4358624               284.2 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap16
BenchmarkSharedMap16-8           5396670               226.6 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap32
BenchmarkSharedMap32-8           6268821               187.0 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap64
BenchmarkSharedMap64-8           6907849               219.6 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap128
BenchmarkSharedMap128-8          6311558               184.3 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap256
BenchmarkSharedMap256-8          6918116               188.7 ns/op            25 B/op          3 allocs/op
BenchmarkSharedMap512
BenchmarkSharedMap512-8          6773053               213.0 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap1024
BenchmarkSharedMap1024-8         5794174               196.6 ns/op            24 B/op          3 allocs/op
PASS
ok      github.com/polarismesh/polaris-server/service/healthcheck       20.769s

**Comment by liu-song** [comment]
上一版本，不能准确计算出 shardmap 中元素个数的benchmark:
```
goos: darwin
goarch: amd64
pkg: github.com/polarismesh/polaris-server/service/healthcheck
cpu: Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
BenchmarkSyncMap
BenchmarkSyncMap-8               1079414              1141 ns/op             133 B/op          4 allocs/op
BenchmarkMutexMap
BenchmarkMutexMap-8              2465654               486.6 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap2
BenchmarkSharedMap2-8            3091468               394.6 ns/op             9 B/op          2 allocs/op
BenchmarkSharedMap4
BenchmarkSharedMap4-8            3797118               322.9 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap8
BenchmarkSharedMap8-8            5420554               228.8 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap16
BenchmarkSharedMap16-8           7416206               168.7 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap32
BenchmarkSharedMap32-8           8286289               143.7 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap64
BenchmarkSharedMap64-8           8924622               147.2 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap128
BenchmarkSharedMap128-8          8192433               143.0 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap256
BenchmarkSharedMap256-8          8961814               139.5 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap512
BenchmarkSharedMap512-8          8772096               138.9 ns/op            24 B/op          3 allocs/op
BenchmarkSharedMap1024
BenchmarkSharedMap1024-8         9686130               135.2 ns/op            24 B/op          3 allocs/op
PASS
ok      github.com/polarismesh/polaris-server/service/healthcheck       21.654s

```

File: service/healthcheck/shardmap.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+		for k, v := range shard.healthCheckInstances {
+			fn(k, v)
+		}
+		shard.healthCheckMutex.RUnlock()
+	}
+}
+
+// Count returns the number of elements within the map.
+func (m *ShardMap) Count() int32 {
+	return m.Len
```
Comment (chuntaojun) [inline service/healthcheck/shardmap.go:152]: atomic.Load(&m.size)

File: service/healthcheck/shardmap.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	"sync"
+	"sync/atomic"
+)
+
+// A Concurrent safe ShardMap for healthCheckInstances
+// To avoid lock bottlenecks this map is dived to several (ShardSize) Concurrent map.
+type ShardMap struct {
+	shardSize uint32
+	Shards    []*Shard
+	Len       int32
```
Comment (chuntaojun) [inline service/healthcheck/shardmap.go:30]: Len -> size

File: service/healthcheck/shardmap.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+package healthcheck
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// A Concurrent safe ShardMap for healthCheckInstances
+// To avoid lock bottlenecks this map is dived to several (ShardSize) Concurrent map.
+type ShardMap struct {
```
Comment (chuntaojun) [inline service/healthcheck/shardmap.go:27]: type shardMap struct {

}

**Comment by chuntaojun** [comment]
新增的注释都改为英文吧

File: service/healthcheck/shardmap.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+// A Concurrent safe ShardMap for healthCheckInstances
+// To avoid lock bottlenecks this map is dived to several (ShardSize) Concurrent map.
+type shardMap struct {
+	shardSize uint32
+	Shards    []*Shard
+	size      int32
+}
+
+type Shard struct {
```
Comment (chuntaojun) [inline service/healthcheck/shardmap.go:33]: Shard => shard

**Review by chuntaojun — APPROVED** [review]

**Pull request merged**